### PR TITLE
ALL: correct MAV_CMD_PREFLIGHT_SET_SENSOR_OFFSETS

### PR DESCRIPTION
**Pull request by khancyr**
use negative value in unsigne int.

### CODE REVIEW

**Comment by OXINARF** [comment]
When you want a commit that affects multiple vehicles/libraries you should use *Global:*, not ALL. In any case, that should be used very rarely and I think it's not appropriate here (there's no reason these changes couldn't be applied separately for all vehicles).

Also, are you aware that these changes don't really matter? Before it was using 255 as the *not set* value, now it uses -1. There's no issue in using a negative value for an unsigned integer, the standard says:

> If the destination type is unsigned, the resulting value is the least unsigned integer congruent to the source integer (modulo 2^n where n is the number of bits used to represent the unsigned type). [ Note: In a two’s complement representation, this conversion is conceptual and there is no change in the bit pattern (if there is no truncation). —end note ]

If you really want to make it better I think the best way is adding an else when checking for *param1* value with a break inside. Then there's no need to check for *compassNumber* value.

**Comment by khancyr** [comment]
@OXINARF Ok for splitting the pr per vehicle.
About the -1, I think it is an error to use it because it is perturbing !

**Comment by OXINARF** [comment]
@khancyr It's not an error and it's quite common to use it.

Like I said, I think the correct way to "fix" this is to check only for known *param1* values and break if none found. If one is found then proceed to update the offsets. This way any update to MAVLink spec (adding 255 or -1 as valid values for example) will not be supported but, at least, correctly respected.

File: APMrover2/GCS_Mavlink.cpp [hunk 1]
```diff
@@ -926,15 +926,15 @@ void GCS_MAVLINK_Rover::handleMessage(mavlink_message_t* msg)
...
+                int8_t compassNumber = -1;
                 if (is_equal(packet.param1, 2.0f)) {
                     compassNumber = 0;
                 } else if (is_equal(packet.param1, 5.0f)) {
                     compassNumber = 1;
                 } else if (is_equal(packet.param1, 6.0f)) {
                     compassNumber = 2;
                 }
-                if (compassNumber != (uint8_t) -1) {
+                if (compassNumber != static_cast<int8_t>(-1)) {
```
Comment (magicrub) [inline APMrover2/GCS_Mavlink.cpp]: why is -1 being cast to static_cast<int8_t>?

**Comment by lucasdemarchi** [comment]
cast from -1 to an unsigned type is a very common way to get the maximum value for that type. If you want to make it cleaner in some places, use <inttypes.h> and then UINT8_MAX, etc.

**Comment by lucasdemarchi** [comment]
Closing this for now... we could make them UINT8_MAX, but the type should remain unsigned.

**Pull request closed without merging**